### PR TITLE
Split NetAttributesExtractor into NetClientAttributesExtractor and NetServerAttributesExtractor

### PR DESCRIPTION
**Pull request by trask**
Alternative to #4278 that implements parallel hierarchies:

* NetClientAttributesExtractor
* InetSocketAddressNetClientAttributesExtractor

and

* NetClientAttributesExtractor
* InetSocketAddressNetClientAttributesExtractor

Note: `peer.service` only applies to client spans ([afaict](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/span-general.md#general-remote-service-attributes)), so `PeerServiceAttributesExtractor` is not split, and only applies to `NetClientAttributesExtractor`.

### CODE REVIEW

**Comment by mateuszrzeszutek** [comment]
Should we go ahead with the `NetAttributesServerExtractor`/`NetAttributesClientExtractor` idea? It matches the recently changes `flavor()` methods in HTTP extractors.

**Comment by trask** [comment]
> Should we go ahead with the `NetAttributesServerExtractor`/`NetAttributesClientExtractor` idea? It matches the recently changes `flavor()` methods in HTTP extractors.

This PR is updated now with this approach 👍

File: instrumentation/jedis/jedis-1.4/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jedis/v1_4/JedisNetAttributesExtractor.java [hunk 1]
```diff
@@ -5,10 +5,10 @@
 
 package io.opentelemetry.javaagent.instrumentation.jedis.v1_4;
 
-import io.opentelemetry.instrumentation.api.instrumenter.net.NetAttributesExtractor;
+import io.opentelemetry.instrumentation.api.instrumenter.net.NetAttributesServerExtractor;
 import org.checkerframework.checker.nullness.qual.Nullable;
 
-final class JedisNetAttributesExtractor extends NetAttributesExtractor<JedisRequest, Void> {
+final class JedisNetAttributesExtractor extends NetAttributesServerExtractor<JedisRequest, Void> {
```
Comment (mateuszrzeszutek) [inline instrumentation/jedis/jedis-1.4/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jedis/v1_4/JedisNetAttributesExtractor.java:11]: Should we use a client extractor here? It's a redis client library after all (even if the net info is available on start)

**Comment by trask** [comment]
> > please see [this commit](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/4287/commits/298f306f1595f97202c6062755c46a434c15d131) for attempt to expose `peerService` to library instrumentation based on discussion in #4235
> 
> I also added similar for Dubbo

and gRPC

File: instrumentation/apache-dubbo-2.7/library/src/main/java/io/opentelemetry/instrumentation/apachedubbo/v2_7/DubboTracingBuilder.java [hunk 2]
```diff
@@ -42,19 +53,37 @@ public DubboTracingBuilder addAttributesExtractor(
...
+          AttributesExtractor.constant(SemanticAttributes.PEER_SERVICE, peerService));
+    } else {
+      clientInstrumenterBuilder.addAttributesExtractor(
+          PeerServiceAttributesExtractor.create(netClientAttributesExtractor));
+    }
 
     return new DubboTracing(
-        builder.newServerInstrumenter(new DubboHeadersGetter()),
-        builder.newClientInstrumenter(new DubboHeadersSetter()));
+        clientInstrumenterBuilder.newServerInstrumenter(new DubboHeadersGetter()),
```
Comment (mateuszrzeszutek) [inline instrumentation/apache-dubbo-2.7/library/src/main/java/io/opentelemetry/instrumentation/apachedubbo/v2_7/DubboTracingBuilder.java:86]:
```suggestion
        serverInstrumenterBuilder.newServerInstrumenter(new DubboHeadersGetter()),
```

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/AttributesExtractor.java [hunk 3]
```diff
@@ -51,4 +51,9 @@ protected abstract void onEnd(
       attributes.put(key, value);
     }
   }
+
+  public static <REQUEST, RESPONSE, T> AttributesExtractor<REQUEST, RESPONSE> constant(
```
Comment (mateuszrzeszutek) [inline instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/AttributesExtractor.java:55]:
```suggestion
  /**
   * Returns an {@link AttributesExtractor} implementation that always extracts the provided constant value.
   */
  public static <REQUEST, RESPONSE, T> AttributesExtractor<REQUEST, RESPONSE> constant(
```

**Review by mateuszrzeszutek — APPROVED** [review]

**Pull request merged**